pkg/trace/obfuscate/credit_cards.go

```diff
@@ -0,0 +1,268 @@
+type ccObfuscator struct {
+	luhn bool
+}
+
+func newCreditCardsObfuscator(useLuhn bool) *ccObfuscator {
+	cco := &ccObfuscator{luhn: useLuhn}
+	pb.SetMetaHook(cco.CheckTag)
+	return cco
+}
+
+// CheckTag checks the tag with the given key and val and returns the final
+// value to be assigned to this tag.
+//
+// For example, in this specific use-case, if the val is detected to be a credit
+// card number, "?" will be returned.
+func (cco *ccObfuscator) CheckTag(k, v string) (newval string) {
+	switch k {
+	case "_sample_rate",
+		"_sampling_priority_v1",
+		"error",
+		"error.msg",
+		"error.type",
+		"error.stack",
+		"env",
+		"graphql.field",
+		"graphql.query",
+		"graphql.type",
+		"graphql.operation.name",
+		"grpc.code",
+		"grpc.method",
+		"grpc.request",
+		"http.status_code",
+		"http.method",
+		"runtime-id",
+		"out.host",
+		"out.port",
+		"sampling.priority",
+		"span.type",
+		"span.name",
+		"service.name",
+		"service",
+		"sql.query",
+		"version":
+		// these tags are known to not be credit card numbers
+		return v
+	}
+	if strings.HasPrefix(k, "_dd") {
+		return v
+	}
+	if isCardNumber(v, cco.luhn) {
+		return "?"
+	}
+	return v
+}
+
+// isSensitve checks if b could be a credit card number by checking the digit count and IIN prefix.
+// If validateLuhn is true, the Luhn checksum is also applied to potential candidates.
+func isCardNumber(b string, validateLuhn bool) (ok bool) {
+	//
+	// Just credit card numbers for now, based on:
+	// • https://baymard.com/checkout-usability/credit-card-patterns
+	// • https://www.regular-expressions.info/creditcard.html
+	//
+	if len(b) == 0 {
+		return false
+	}
+	if b[0] != 0x20 && b[0] != 0x2d && (b[0] < 0x30 || b[0] > 0x39) {
+		// fast path: only valid characters are 0-9, space (" ") and dash("-")
+		return false
+	}
+	i := 0                      // byte index for traversing b
+	prefix := 0                 // holds up to b[:6] digits as a numeric value (for example []byte{"523"} becomes int(523)) for checking prefixes
+	count := 0                  // counts digits encountered
+	foundPrefix := false        // reports whether we've detected a valid prefix
+	recdigit := func(_ byte) {} // callback on each found digit; no-op by default (we only need this for Luhn)
+	if validateLuhn {
+		// we need Luhn checksum validation, so we have to take additional action
+		// and record all digits found
+		var buf strings.Builder
+		recdigit = func(b byte) { buf.WriteByte(b) }
+		defer func() {
+			if !ok {
+				// if isCardNumber returned false, it means that b can not be
+				// a credit card number
+				return
+			}
+			// potentially a credit card number, run the Luhn checksum
+			ok = luhnValid(buf.String())
+		}()
+	}
+loop:
+	for i < len(b) {
+		// We traverse and search b for a valid IIN credit card prefix based
+		// on the digits found, ignoring spaces and dashes.
+		// Source: https://www.regular-expressions.info/creditcard.html
+		switch b[i] {
+		case 0x20, 0x2d:
```

Comment:
💅 Using rune literals here would allow you to avoid the comment, same with other `0x` encoded characters elsewhere. Shouldn't have any perf overhead as it gets sorted out at compile time.

```suggestion
		case case ' ', '-':
```

Reply:
Right. Replaced throughout!